contributing.md

```diff
@@ -0,0 +1,45 @@
+# Contributing
+
+## Installing Dependencies
+
+Sphinxcontrib-Needs requires only [Poetry](https://python-poetry.org/) to be installed as a system dependency, the rest of the dependencies are installed in an isolated environment by Poetry.
+
+1. [Install Poetry](https://python-poetry.org/docs/#installation)
+2. Install project dependencies
+   
+        poetry install
+
+
+## Running Tests
+
+in order to run the tests-
+
+    poetry run nosetests -w tests
+
+a Makefile is provided
+
+    make test
+
+## Running Test Matrix
+
+This project provides a test matrix for running the tests across a range of python and sphinx versions.
+
+[Nox](https://nox.thea.codes/en/stable/) is used as a test runner.
+
+Running the matrix tests requires additional system-wide dependencies
+
+1. [Install Nox](https://nox.thea.codes/en/stable/tutorial.html#installation)
```

Comment:
it *can* be a normal test dependency, but it's better if it's a system dependency. there's a couple of reasons for this

- most contributors won't need to install or use Nox *at all* to work with this project. they can use flake8, nose, etc within the poetry environment, and leave the CI job to run the full test matrix using Nox. to reiterate, Nox is *not* required for running tests. It is *only* required for running the full test matrix across different interpreter and dependency versions
- all of the dev-dependencies are installed in a single python environment, meaning they must all be compatible with each other. Nox introduces a lot of extra dependencies to that environment that constrain Poetry's ability to find a compatible combination. In particular, you are supporting some very old dependencies in this project, so you need to be careful about introducing new constraints like this. Since most contributors don't need Nox, they shouldn't need to pay this 'cost'.

Reply:
ah i remember why. it's because the test dependencies are a subset of the `dev-dependencies`. it's not necessary to install all of the dev dependencies. this creates a significant time saving if running the full test matrix sequentially.

the difference is even more striking when you're doing something like linting, for which only 1 or 2 of the dev-dependencies is required.

a neater solution would be to group the dev dependencies (keep an eye on https://github.com/python-poetry/poetry/issues/1644)

Reply:
I understand, good points.